### ISSUE TITLE
Huge performance loss with adding values to list

### ISSUE THREAD

**Issue by JeepSmash** [bensku/Skript#891]
I noticed an incredible amount of performance loss when trying to create some custom world editing commands.  The problem become extremely evident when making larger edits (radius greater than 5).

The super slow way:
```
loop blocks in radius {_rad} around location of player:
  add loop-block to {blocks::*}
```

The 100+ times faster way:
```
set {_index} to 0
loop blocks in radius {_rad} around location of player:
  set {blocks::%{_index}%} to loop-block
  add 1 to {_index}
```

I found that the first way could only process about 100 blocks in the time the second way can process about 10000.  I'm not sure why such a drastic difference is coming but i felt it was significant enough to report here as a possible issue.  (or at least help others having lag issues with the workaround I found to drastically help)  My thoughts are that the lag is coming from searching the database for the next usable index when adding a value to a list since that is really the only thing different between the two except the second example actually has more lines of code to process by runs 100 times faster literally.

**Comment by bensku** [comment]
Hmm, I think Skript uses TreeMaps under the hood. Getting size of those is somewhat expensive.

**Comment by Pikachu920** [comment]
I assume `set {_blocks::*} to blocks in radius {_rad} of player` has the same result as the slow method?

**Comment by TrademarkTM** [comment]
Well, I'm going to leave a dot here because it really needs to be fixed
.
Can't you just save the last index of the variable?

**Comment by bensku** [comment]
Yes, but that requires something that is not a pile of recursive TreeMaps (which is how lists are currently implemented). So not that easy to do, sadly.

**Comment by bi0qaw** [comment]
I think the problem is that Skript iterates over the whole list every time you add a value to it. The reason is that it looks for missing indices and tries to add the new value there. Example script:
```
set {_list::1} to "value 1"
set {_list::2} to "value 2"
set {_list::3} to "value 3"
clear {_list::2}
add "new value" to {_list::*}
broadcast "%{_list::*}%"
```
Results in:
```
value 1, new value and value 3
``` 

The new value gets added to index 2 because it is the first one that has no value. While this seems like a nice feature it is pretty slow and doesn't make much sense. Especially because you can have keys with negative values, keys that are not numbers... Furthermore, if you remove a value, the list is not reindexed. There is no easy fix to this if we don't want to break backwards compatibility. 

One possible solution is to check whether there is an index missing in the current list and only then loop through it. If there is no index missing we are fine and just set the new value at index: `size + 1`. Checking if an index is missing is pretty simple. We just have to compare the size of the list with the lastKey() in the TreeMap. If they are equal we know that all the indices are in order. If they are not equal we have either non-integer indices or indices are missing. In that case we would have to loop through all the keys.

@bensku I think we could just add the check [here](https://github.com/bensku/Skript/blob/6e0f9dc81703cc175bc4bbf88435671c4acf9f7d/src/main/java/ch/njol/skript/lang/Variable.java#L490)

**Comment by Liicra** [comment]
when this will be fixed?

**Comment by TPGamesNL** [comment]
Here's a JFR result from this code:
```
import:
	java.lang.System

command /test:
	trigger:
		set {_start} to System.currentTimeMillis()
		set {_amount} to 4000
		loop {_amount} times:
			add 1 to {_l::*}
		set {_end} to System.currentTimeMillis()
		broadcast "Adding took %{_end} - {_start}%ms for %{_amount}% times"
```
![image](https://user-images.githubusercontent.com/29547183/139585339-f50a9052-a600-45e6-913a-ae29f2a8814a.png)
(both start and end of JFR where during loop, loop itself took 12 seconds, recording took 9 seconds, 371 total samples)

**Comment by MinecraftMan1013number2** [comment]
Any update on this SkriptLang team?

**Comment by bluelhf** [comment]
I'm no SkriptLang team member, but adding to lists iterates over the list to look for the right numeric index to add to. This means that adding to lists is done in O(n) time, whereas setting indices is done in O(1) time — this is expected behaviour.

If you're looking for high performance, use the other method of adding to lists.

P.s. Skript is not meant to be _fast_, if you want to write performant code, you should probably write it in Java :)

**Comment by Moderocky** [comment]
> Any update on this SkriptLang team?

I have given them a way to fix it, so it’s just a matter of waiting for somebody to PR it :)

**Comment by EysseW** [comment]
Wasted hours on this bug. Needs to be fixed asap. I honestly don't get why it has a "medium" priority.

**Comment by Fusezion** [comment]
> Wasted hours on this bug. Needs to be fixed asap. I honestly don't get why it has a "medium" priority.

Falls under "medium" as it's just a lack of proper implementation, this isn't a bug as much as just poor design.
